src/timeline/RoomlistModel.h

```diff
@@ -41,6 +48,51 @@ class RoomPreview
     bool isInvite_ = false;
 };
 
+#ifdef NHEKO_DBUS_SYS
+class RoomlistModel;
+
+class RoomListDBusInterface : public QObject
+{
+    Q_OBJECT
+
+public:
+    RoomListDBusInterface(RoomlistModel *parent);
+
+public slots:
+    //! Get the nheko D-Bus API version.
+    Q_SCRIPTABLE QString apiVersion() const { return nheko::dbus::apiVersion; }
+    //! Call this function to get a list of all joined rooms.
+    Q_SCRIPTABLE QVector<nheko::dbus::RoomInfoItem> getRooms(const QDBusMessage &message);
+    //! Activates a currently joined room.
+    Q_SCRIPTABLE void activateRoom(const QString &roomid) const;
+    //! Joins a room. It is your responsibility to ask for confirmation (if desired).
+    Q_SCRIPTABLE void joinRoom(const QString &roomid) const;
+
+private slots:
+    void prepareModel();
+
+private:
+    void bringWindowToTop() const;
+
+    RoomlistModel *m_parent;
+
+    // this is a QSharedPointer so that copy ops are less expensive (see below)
+    QSharedPointer<QVector<nheko::dbus::RoomInfoItem>> m_model{
```

Comment:
I'm not sure I am fond of storing all the rooms in memory multiple times... It doesn't scale to accounts like mine :D

Reply:
IIRC, I clear the staging area once it has been saved to the main model.

Reply:
The problem is that dynamically building the roomlist before sending can take a second or two, which is not terribly great. That is mainly due to downloading the images though. What if I would cache images in TimelineModels to remove the lag?